machine/v1beta1/types_awsplacementgroup.go

```diff
@@ -0,0 +1,164 @@
+	AWSSpreadPlacementGroupType AWSPlacementGroupType = "Spread"
+)
+
+// AWSPlacementGroupUnion is a discriminated union of placement group configuration.
+// +union
+type AWSPlacementGroupUnion struct {
+	// GroupType specifies the type of AWS placement group to use for this Machine.
+	// This parameter is only used when a Machine is being created and the named
+	// placement group does not exist.
+	// Valid values are "Cluster", "Partition", "Spread".
+	// This value is required and, in case a placement group already exists, will be
+	// validated against the existing placement group.
+	// Note: If the value of this field is "Spread", Machines created within the group
+	// may no have placement.tenancy set
+	// to "dedicated".
+	// +kubebuilder:validation:Enum:="Cluster";"Partition";"Spread"
+	// +unionDiscriminator
+	// +kubebuilder:validation:Required
+	// +required
+	GroupType AWSPlacementGroupType `json:"groupType,omitempty"`
+
+	// +optional
+	Partition *AWSPartitionPlacement `json:"partition,omitempty"`
+}
+
+// AWSPartitionPlacement defines the configuration for partition placement groups.
+type AWSPartitionPlacement struct {
+	// Count specifies the number of partitions for a Partition placement
+	// group. This value is only observed when creating a placement group and
+	// only when the `groupType` is set to `Partition`.
+	// Note the partition count of a placement group cannot be changed after creation.
+	// If unset, AWS will provide a default partition count.
+	// This default is currently 2.
+	// Note: When using more than 2 partitions, placement.tenancy "dedicated" on Machines
+	// created within the group is unavailable.
+	// +kubebuilder:validation:Minimum:=1
+	// +kubebuilder:validation:Maximum:=7
+	// +optional
+	Count int32 `json:"count,omitempty"`
+}
+
+type AWSPlacementGroupStatus struct {
+	// LastUpdated identifies when this status was last observed.
+	// +optional
+	LastUpdated *metav1.Time `json:"lastUpdated,omitempty"`
+
+	// Replicas counts how many AWS EC2 instances are present within the placement group.
+	// Note: This is a pointer to be able to distinguish between an empty placement group
+	// and the status having not yet been observed.
+	// +optional
+	Replicas *int32 `json:"replicas,omitempty"`
+
+	// ManagementState determines whether the placement group is expected
+	// to be managed by this CRD or whether it is user managed.
+	// A managed placement group may be moved to unmanaged, however an unmanaged
+	// group may not be moved back to managed.
+	// This value is owned by the controller and may differ from the spec in cases
+	// when a user attempts to manage a previously unmanaged placement group.
+	// +optional
+	ManagementState ManagementState `json:"managementState,omitempty"`
+
+	// ObservedConfiguration represents the configuration present on the placement group on AWS.
+	// +optional
+	ObservedConfiguration AWSPlacementGroupUnion `json:"observedConfiguration,omitempty"`
```

Comment:
I suggest not re-using this type.  I won't permanently stick on it, but often spec and status drift from each other and I would keep the separate from the beginning so you don't accidentally add fields that are not honored or not set.